### PR TITLE
Converting between different physical units [User Story #43]

### PR DESCRIPTION
**Pull request by izconcept**
User Story #43 
Tasks #134 #135 #136

## WHAT kind of change does this PR introduce?

Added a new view which allows the conversion of three different physical units. Weight, length, and volume. 

The supported units are:

### Weight
* Grams (g)
* Kilograms (kg)
* Ounces (oz)
* Pounds (lb)

### Length
* Centimeters (cm)
* Meters (m)
* Inches (in)
* Feet (gt)

### Volume
* Millilitres (ml)
* Liters (l)
* Quart (qt)
* Gallon (gl)

Furthermore, this PR also contains a small architectural redesign of the implementation classes. Which reduces code reuse and falls in line with good OOP practices.

### The new Activity
A new view/layout is used. The new view resembles `MoneyActivity` but makes some important changes to the inner workings. Crucially, some of the helper methods in the `BaseCalculatorImpl` were overwritten to support unlimited decimal numbers (`MoneyCalcultorImpl` only had support for 2 decimal places)

Three new operation classes were defined that allowed us to do weight, length and volume conversions.

New integration tests were implemented in `UnitConversionActivityTest`

New unit tests were implemented in `ConversionCalculatorImplTest`

### The helper refactor

Since the impl helper for `UnitConversionActivity` and the helper for the `MoneyActivity` share a lot of the same code, the overlapping code was put into a parent class called `BaseCalculatorImpl` from which `MoneyCalculatorImp` and `ConversionCalculatorImpl` is now inheriting from. I believe this design choice falls in line with good OOP practices.

## Checklist
- [ ] I have reviewed and tested the changes :white_check_mark:
- [ ] I have squashed my commits to have a reasonable amount of commits
- [ ] All of the code I have written adhere to the coding conventions outlined in the coding conventions section of the wiki

### CODE REVIEW

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/UnitConversionActivity.kt [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        fromBuilder.setPositiveButton("Convert", DialogInterface.OnClickListener { dialog, id ->
+            val from = fromSpinner.selectedItem.toString()
+            val to = toSpinner.selectedItem.toString()
+            when (type) {
+                "weight" -> calc.performWeightConversion(from, to)
+                "length" -> calc.performLengthConversion(from, to)
+                else -> calc.performVolumeOperation(from, to)
+            }
+        })
+                .setNegativeButton("Cancel", DialogInterface.OnClickListener { dialog, id ->
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/activities/UnitConversionActivity.kt]: I think you could just rid of the lambda here.

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/BaseCalculatorImpl.kt [hunk 2]
```diff
@@ -0,0 +1,89 @@
+package com.simplemobiletools.calculator.helpers
+
+import android.content.Context
+import com.simpletools.calculator.commons.R
+import com.simpletools.calculator.commons.helpers.Calculator
+import com.simpletools.calculator.commons.helpers.Formatter
+
+open class BaseCalculatorImpl(calculator: Calculator, open val context: Context) {
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/helpers/BaseCalculatorImpl.kt]: what does open mean here

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/MoneyCalculatorImpl.kt [hunk 3]
```diff
@@ -3,100 +3,18 @@ package com.simplemobiletools.calculator.helpers
...
 
-class MoneyCalculatorImpl(calculator: Calculator, val context: Context) {
+class MoneyCalculatorImpl(calculator: Calculator, override val context: Context) : BaseCalculatorImpl(calculator, context) {
     private var mCallback: Calculator? = calculator
-    private var number: String = ""
-    private var decimalClicked: Boolean = false
-    private var decimalCounter: Int = 0
     private var taskBuilder: BackgroundCurrencyTaskBuilder = BackgroundCurrencyTaskBuilder(calculator, this)
 
-    init {
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/helpers/MoneyCalculatorImpl.kt]: Nice refactoring is always good

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -43,6 +44,26 @@
         <item>CAD</item>
     </string-array>
 
+    <string-array name="weight">
```
Comment (almiche) [inline app/src/main/res/values/strings.xml]: Im not too sure on this but I believe a more appropriate place for these would be in arrays.xml. Up to you, either works.

File: commons/src/main/java/com/simpletools/calculator/commons/operations/LengthOperation.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package com.simpletools.calculator.commons.operations
+
+object LengthOperation {
+    var value: Double = 0.0
+    var conversion: String = ""
+    var LengthMap = HashMap<String, Double>()
+
+    init {
+        LengthMap.put("cm cm", 1.0)
```
Comment (almiche) [inline commons/src/main/java/com/simpletools/calculator/commons/operations/LengthOperation.kt]: why not use underscores as a separator ? Also I wonder if this can be put inside of a json file to be more easily maintained. The JSON could contain all three of the conversionMaps. Eliminating the need for three classes.

**Comment by almiche** [comment]
✅ Styling is adequate and follows coding conventions outlined by the [Kotlin Foundation](https://kotlinlang.org/docs/reference/coding-conventions.html)

✅ New methods and functionalities have appropriate tests (testing edge cases + good coverage)

**Testing is through**

✅ Branch has an acceptable amount of commits with descriptive names and

✅   PR description matches the code outlined a has mentions to issues

✅ This PR attempts to implement a specific feature (no code blob or epic pr)

Status: **Requested Changes**

File: .travis.yml [hunk 6]
```diff
@@ -33,7 +33,7 @@ jobs:
         - ./gradlew compileDebugUnitTestSources
         - ./gradlew compileReleaseSources
     - stage: Testing (Parallel integration & unit tests) 🏎
-      script: ./testing integration_tests
+      script: travis_wait 30 ./testing integration_tests
```
[the remark this answers is not in the document]
Comment (izconcept) [reply .travis.yml]: Tests were failing due to the timeout and doing a re-base at the time may have introduced conflicts

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/UnitConversionActivity.kt [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        val myList = ColorStateList(states, colors)
+
+        getUnitConversionButtonIds().forEach {
+            it.setTextColor(getContrastColor(color))
+            ViewCompat.setBackgroundTintList(it, myList)
+        }
+    }
+
+    @SuppressLint("Range")
+    private fun getContrastColor(color: Int): Int {
```
[the remark this answers is not in the document]
Comment (izconcept) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/UnitConversionActivity.kt]: Didn't wan't to restructure all the activities to inherit from a parent activity since this would have caused conflicts with a lot of the other PRs (the crypto and bitwise). We can consider doing this once all of them have been merged.

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/BaseCalculatorImpl.kt (hunk 2 above)
Comment (izconcept) [reply app/src/main/kotlin/com/simplemobiletools/calculator/helpers/BaseCalculatorImpl.kt]: It's used to declare an explicit supertype. This class is never instantiated. It's being inherited by two other helpers.

File: app/src/main/res/values/strings.xml (hunk 4 above)
Comment (izconcept) [reply app/src/main/res/values/strings.xml]: The list for the currency converter spinner was also here so I put mine here as well.

File: commons/src/main/java/com/simpletools/calculator/commons/operations/LengthOperation.kt (hunk 5 above)
Comment (izconcept) [reply commons/src/main/java/com/simpletools/calculator/commons/operations/LengthOperation.kt]: That would reduce the cohesion of the class(es) and would violate our current operation 'design pattern' where operations are classes that perform a single thing.

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/UnitConversionActivity.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/UnitConversionActivity.kt]: Indeed there is a lot of duplicate code that is shared among activities so once all the features have been merged we should refactor them.

File: app/src/main/res/values/strings.xml (hunk 4 above)
Comment (miisaka) [reply app/src/main/res/values/strings.xml]: As per [Android's documentation](https://developer.android.com/guide/topics/resources/string-resource.html#StringArray), there is no need to create a new arrays.xml file to store String Arrays as they are still String ressources.

File: commons/src/main/java/com/simpletools/calculator/commons/operations/LengthOperation.kt [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        LengthMap.put("in m", 0.0254)
+        LengthMap.put("in in", 1.0)
+        LengthMap.put("in ft", 0.0833333)
+        LengthMap.put("ft cm", 30.48)
+        LengthMap.put("ft m", 0.3048)
+        LengthMap.put("ft in", 12.0)
+        LengthMap.put("ft ft", 1.0)
+    }
+
+    fun setParams(value: Double, conversion: String): LengthOperation? {
```
Comment (miisaka) [inline commons/src/main/java/com/simpletools/calculator/commons/operations/LengthOperation.kt]: You should consider making these methods as well as variables private in order to enforce data encapsulation.

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/UnitConversionActivity.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (almiche) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/UnitConversionActivity.kt]: Fair enough

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/BaseCalculatorImpl.kt (hunk 2 above)
Comment (almiche) [reply app/src/main/kotlin/com/simplemobiletools/calculator/helpers/BaseCalculatorImpl.kt]: I see

**Comment by almiche** [comment]
✅ Styling is adequate and follows coding conventions outlined by the [Kotlin Foundation](https://kotlinlang.org/docs/reference/coding-conventions.html)

✅ New methods and functionalities have appropriate tests (testing edge cases + good coverage)

✅ Branch has an acceptable amount of commits with descriptive names and

✅   PR description matches the code outlined a has mentions to issues

✅ This PR attempts to implement a specific feature (no code blob or epic pr)

Status: **Approved**
